File: backend/app/models/cultural_token.rb
# Represents a cultural element extracted from a niche with design potential.
# Cultural tokens are key symbols, phrases, or concepts that resonate with
# a community and can be transformed into marketable designs.
#
# @attr [Integer] id Primary key
# @attr [Integer] niche_id Foreign key to Niche
# @attr [String] token_type Type of token (e.g., 'phrase', 'symbol', 'character', 'meme')
# @attr [String] value The actual content of the token
# @attr [Text] context Additional context explaining the token's cultural significance
# @attr [Float] composite_score Overall viability score for design conversion
# @attr [String] status Current production status
# @attr [DateTime] created_at Record creation timestamp
# @attr [DateTime] updated_at Record update timestamp
class CulturalToken < ApplicationRecord
  belongs_to :niche
  has_many :designs, dependent: :destroy

  validates :token_type, presence: true
  validates :value, presence: true
  validates :status, presence: true

  # Production pipeline status for token-to-design conversion
  # @note Default status is :extracted
  enum :status, {
    extracted: "extracted",
    designs_pending: "designs_pending",
    designs_ready: "designs_ready",
    in_production: "in_production",
    listed: "listed"
  }, default: :extracted

  # @!scope class
  # Returns tokens sorted by composite score in descending order
  # @return [ActiveRecord::Relation<CulturalToken>] Tokens ordered by design viability
  scope :by_composite_score, -> { order(composite_score: :desc) }

  # @!scope class
  # Returns tokens ready for design generation
  # @return [ActiveRecord::Relation<CulturalToken>] Tokens with extracted status
  scope :ready_for_designs, -> { where(status: :extracted) }
end

File: frontend/src/types/cultural-token.ts
export const TOKEN_TYPES = ['quote', 'symbol', 'character', 'meme'] as const
export type TokenType = (typeof TOKEN_TYPES)[number]

export const TOKEN_STATUSES = [
  'extracted',
  'designs_pending',
  'designs_ready',
  'in_production',
  'listed',
] as const
export type TokenStatus = (typeof TOKEN_STATUSES)[number]

export const STATUS_LABELS: Record<TokenStatus, string> = {
  extracted: 'Extracted',
  designs_pending: 'Designs Pending',
  designs_ready: 'Designs Ready',
  in_production: 'In Production',
  listed: 'Listed',
}

export interface CulturalTokenFull {
  id: number
  niche_id: number
  token_type: TokenType
  value: string
  status: TokenStatus
  composite_score: number
  frequency_score: number
  emotional_intensity: number
  visual_potential: number
  uniqueness_score: number
  source_references: unknown
  context: string
  created_at: string
  updated_at: string
}

export type SortColumn =
  | 'value'
  | 'token_type'
  | 'frequency_score'
  | 'emotional_intensity'
  | 'visual_potential'
  | 'uniqueness_score'
  | 'composite_score'
  | 'status'

export type SortDirection = 'ASC' | 'DESC'
